### PR TITLE
Multithreaded training

### PR DESCRIPTION
**Pull request by tmbo**
- use multiple threads during training
- new option `-t` to specify number of threads
- spacy will only use multiple threads during intent classification - not during entity training
- small fix to save mitie models in the same way as spacy (added `create_unique_subfolder` option)

### CODE REVIEW

**Comment by amn41** [comment]
just a hint on the failing tests: there are 2 pep-8 issues & you have to update the default config to include the `num_threads` (see here : https://github.com/golastmile/rasa_nlu/blob/master/src/config.py#L11 )

**Comment by tmbo** [comment]
good point, fixed them.

**Comment by amn41** [comment]
code looks good. I added some tests and travis is erroring out from lack of memory. We actually have the same issue on the multitenancy branch. 

I wonder if we can control the test execution, ensuring they get run in series and with proper garbage collection in between.

**Comment by tmbo** [comment]
That is strange, I think the tests are executed sequentially anyway and the garbage collection should kick in automatically as well. Maybe spacy / mitie are doing something here, do we need to "unload" them?

**Comment by tmbo** [comment]
@amn41 Alright, seems we are good to go now. Additional changes:

Some travis improvements:
- cache `pip`
- run test boxed (spawns one interpreter per test --> ensures memory is freed up)

Test changes:
- use temp files --> allows to run boxed tests in parallel

File: _pytest/trainers/test_mitie_trainer.py [hunk 1]
```diff
@@ -7,4 +7,4 @@ class TestMitieTrainer:
     def test_failure_on_invalid_lang(self):
         config = RasaNLUConfig("config_mitie.json")
         with pytest.raises(NotImplementedError):
-            MITIETrainer(config.mitie_file, 'umpalumpa')
+            MITIETrainer(config['mitie_file'], 'umpalumpa')
```
Comment (amn41) [inline _pytest/trainers/test_mitie_trainer.py]: shouldn't `mitie_file` be set as an attribute? why the change?

Comment (tmbo) [inline _pytest/trainers/test_mitie_trainer.py]: yes `.mitie_file` works as well.

Nevertheless, the dot notation raises a warning in IDEs `mitie_file might not be defined`, whereas the other notation makes clear that it is some sort of dictionary (where values might not be defined)
